LIU-505: Fix multi-input skipped AppDROP after Branch False condition.

myxie opened the pull request:
This is technically only a 1 line fix, but I wanted to start having more branch tests.

# JIRA Ticket 
<!---
If there is no JIRA ticket, please consider raising one to summarise the work there. Alternatively, link to a GitHub if that exists._
--->
LIU-505

# Type
<!---Select what type of work this PR is addressing. This is useful for preparing the [Changelog](https://github.com/ICRAR/daliuge/blob/master/CHANGELOG.md)--->

- [ ] Feature (addition)
- [x] Bug fix
- [ ] Refactor (change)
- [ ] Documentation 

# Problem/Issue 
<!--- Provide an overview of what this PR address--->

Omar has raised a potential bug in how we manage SKIPPED AppDROPS with multiple inputs.

Currently, if I have a PyFuncApp that takes 2 inputs: 1 that is 'Complete', and the other, which is 'Skipped',  the following happens: 

![image](https://github.com/user-attachments/assets/6fd3ae32-515d-4ac8-b598-2d3c22fb2c7e)

This is because of the following code: 

https://github.com/ICRAR/daliuge/blob/98711e70b7e94273daaf323c204e80b175b1824e/daliuge-engine/dlg/apps/app_base.py#L527-L528

Where we require _all_ inputs to be skipped. What happens is that the AppDROP continues thinking that the skipped memory DROP is fine when it has been skipped, and the error occurs.

# Solution
<!--- A description of the solution, including design decisions or compromises made. Consider adding a figure or example of how the behaviour has changed. ---> 

I think that this can be fixed by a one line change: 

https://github.com/ICRAR/daliuge/blob/45bf24a3d91353afc595fa52981551fd7e93f92c/daliuge-engine/dlg/apps/app_base.py#L527-L528

Where we skip any DROP that has at least one input skipped, regardless of how many inputs it has. 

# Checklist
<!--- Provide a description of documentation added, testing, including manual and programmatic ---> 

- [x] Unittests added
    - Note: there is a reciprocal PR for the test graphs: https://github.com/ICRAR/EAGLE_test_repo/pull/18
- ~[ ] Documentation added~
    - Reason for not adding documentation: There is another PR (#347) that will be adding documentation for Graph Patterns, including the Branch. 

## Summary by Sourcery

Fix branch skipping logic for multi-input AppDROP and add comprehensive branch skip tests using new end-to-end utilities

New Features:
- Add end-to-end test utilities for graph translation and execution
- Introduce new branch skipping tests covering single and multiple-input scenarios

Bug Fixes:
- Ensure AppDROP skips execution if any input is skipped by adjusting the skip condition
- Correct minor test assertion formatting in test_drop.py

Enhancements:
- Update test dependency to use the LIU-505 branch of eagle_test_graphs from GitHub
- Remove redundant code in create_and_run_graph_spec_from_graph_file and streamline test utilities

CI:
- Include the new branch tests in the GitHub Actions unit test workflow

## Summary by Sourcery

Fix branch skip logic for multi-input AppDROP and strengthen testing by adding end-to-end utilities and comprehensive branch skip tests, while updating dependencies and CI to support the new test suite.

New Features:
- Introduce end-to-end graph translation and execution utilities for testing.

Bug Fixes:
- Skip AppDROP execution when any of its inputs is skipped, instead of requiring all inputs to be skipped.

Enhancements:
- Refactor graph-spec utilities to centralize and streamline test setup.
- Point eagle_test_graphs dependency to the LIU-505 branch in both engine and translator.

Build:
- Update setup.py in daliuge-engine and daliuge-translator to fetch eagle_test_graphs from GitHub.

CI:
- Include the new branch skip tests in the GitHub Actions unit test workflow.

Tests:
- Add branch skip tests for single-input and multiple-input scenarios.
- Fix minor assertion formatting in existing drop tests.

myxie commented on daliuge-engine/test/dlg_end_to_end_utils.py:
```diff
@@ -65,7 +65,8 @@ def create_and_run_graph_spec(test_case: unittest.TestCase, appDropSpec: list[di
         DROPStates.COMPLETED, DROPStates.SKIPPED, DROPStates.ERROR, DROPStates.CANCELLED]):
         for drop in roots:
-            if app_root:
+            if drop.type != "Data":
```
Was there any reason for modifying this? It's unclear what this is doing differently.

awicenec replied:
This enables to have multiple mixed root nodes, which is the case in the graph I've added to the tests now.

myxie commented on daliuge-engine/dlg/named_port_utils.py:
```diff
@@ -112,6 +113,7 @@ def identify_named_ports(
     keywordArgs: dict,
     check_len: int = 0,
     mode: str = "inputs",
+    skip_on_input=True,
```
I think this is redundant now, as it is not used.

myxie commented on daliuge-engine/dlg/apps/app_base.py:
```diff
@@ -612,7 +616,24 @@ def _run(self):
+            else:
+                ni = named_inputs[attr_name]
+            logger.debug("Identified input: %s", ni.name)
+            if  "componentParams" not in ni.parameters or ni.parameters["componentParams"]["dropclass"]["value"] != "dlg.data.drops.data_base.NullDROP" and ni.status == DROPStates.COMPLETED:
```
I think we need to add further documentation to this method to explain what is going on here. I don't quite understand why we need to check the parsing here?

myxie commented on daliuge-engine/dlg/apps/app_base.py:
```diff
@@ -612,7 +616,24 @@ def _run(self):
+                if not hasattr(self, attr_name):
+                    self.__setattr__(attr_name, named_inputs[attr_name])
+                else: # temporary hack
+                    self.__setattr__(attr_name, load_pickle(ni))
```
We default to `dill` now, should this be load_dill?

myxie commented on daliuge-engine/dlg/apps/pyfunc.py:
```diff
@@ -667,17 +669,21 @@ def _ports2args(self, pargsDict, keyargsDict) -> dict:
                 key = list(inport.keys())[0]
                 inputs_dict[key] = {"name": inport[key], "path": None, "drop":
                     self._inputs[key]}
+            skip_on_input = True
```
I don't believe this is necessary any more, as we don't use it in the `identify_named_ports` function.

myxie commented on daliuge-engine/dlg/drop.py:
```diff
@@ -427,28 +427,26 @@ def get_param_value(attr_name, default_value):
+            # c_param = a_param = default_value
             if has_component_param:
-                param = kwargs.get(attr_name)
+                c_param = kwargs.get(attr_name)
             elif has_app_param:
```
May we call these `component_param`, `attribute_param`? It's just that much more readable.

myxie commented on daliuge-engine/dlg/drop.py:
```diff
@@ -427,28 +427,26 @@ def get_param_value(attr_name, default_value):
-            return param
+                    a_param = kwargs["applicationArgs"].get(attr_name).value
+            if has_component_param:
+                # logger.warning(
```
There's quite a few comments that exist that we can delete now that testing is done.

myxie replied:
This is the second check for `has_component_param`. We don't need the variables `c_param`, `a_param`, we should be able to return `kwargs.get(attr_name)` immediately.

myxie commented on daliuge-engine/dlg/drop.py:
```diff
@@ -427,28 +427,26 @@ def get_param_value(attr_name, default_value):
+            elif has_app_param:
+                return a_param
+            else:
+                return default_value
```
Aren't we having some scope creep between our definitions of component vs attribute here?

awicenec replied:
that is legacy and the commented part was there before. The issue is that in the Apps very often the parameters are set using dlg_*_param functions directly into class attributes and those seem to end up as component_parameters (I don't quite know how, though). Might be worthwhile to have another look and make sure they are put in the correct category. I found it always quite annoying that we always had these log-messages about duplicate parameters.

myxie reviewed (commented):
Hi @awicenec, I've just done a review of the code and the functionality. Thank you very much for picking up this work, the end result is great.

awicenec commented:
It supports graphs with mixed roots, else the new graph I've put in the
test repo would not work.

On Thu, 10 July 2025, 16:19 Ryan Bunney, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
>
> Hi @awicenec <https://github.com/awicenec>, I've just done a review of
> the code and the functionality. Thank you very much for picking up this
> work, the end result is great.
> ------------------------------
>
> In daliuge-engine/dlg/named_port_utils.py
> <https://github.com/ICRAR/daliuge/pull/346#discussion_r2196933333>:
>
> > @@ -112,6 +113,7 @@ def identify_named_ports(
>      keywordArgs: dict,
>      check_len: int = 0,
>      mode: str = "inputs",
> +    skip_on_input=True,
>
> I think this is redundant now, as it is not used.
> ------------------------------
>
> In daliuge-engine/dlg/apps/app_base.py
> <https://github.com/ICRAR/daliuge/pull/346#discussion_r2196943024>:
>
> > @@ -612,7 +616,24 @@ def _run(self):
>              #         attr_name, load_pickle(named_inputs[attr_name])
>              #     )
>              # else:
> -            self.__setattr__(attr_name, named_inputs[attr_name])
> +            if isinstance(named_inputs[attr_name], list) and len(named_inputs[attr_name]) > 1:
> +                for ni in named_inputs[attr_name]:
> +                    if ni.status != DROPStates.COMPLETED:
> +                        continue
> +                    break # we use the first completed
> +            else:
> +                ni = named_inputs[attr_name]
> +            logger.debug("Identified input: %s", ni.name)
> +            if  "componentParams" not in ni.parameters or ni.parameters["componentParams"]["dropclass"]["value"] != "dlg.data.drops.data_base.NullDROP" and ni.status == DROPStates.COMPLETED:
>
> I think we need to add further documentation to this method to explain
> what is going on here. I don't quite understand why we need to check the
> parsing here?
> ------------------------------
>
> In daliuge-engine/dlg/apps/app_base.py
> <https://github.com/ICRAR/daliuge/pull/346#discussion_r2196943751>:
>
> > @@ -612,7 +616,24 @@ def _run(self):
>              #         attr_name, load_pickle(named_inputs[attr_name])
>              #     )
>              # else:
> -            self.__setattr__(attr_name, named_inputs[attr_name])
> +            if isinstance(named_inputs[attr_name], list) and len(named_inputs[attr_name]) > 1:
> +                for ni in named_inputs[attr_name]:
> +                    if ni.status != DROPStates.COMPLETED:
> +                        continue
> +                    break # we use the first completed
> +            else:
> +                ni = named_inputs[attr_name]
> +            logger.debug("Identified input: %s", ni.name)
> +            if  "componentParams" not in ni.parameters or ni.parameters["componentParams"]["dropclass"]["value"] != "dlg.data.drops.data_base.NullDROP" and ni.status == DROPStates.COMPLETED:
> +                # TODO: need to check for parser before reading
> +                if not hasattr(self, attr_name):
> +                    self.__setattr__(attr_name, named_inputs[attr_name])
> +                else: # temporary hack
> +                    self.__setattr__(attr_name, load_pickle(ni))
>
> We default to dill now, should this be load_dill?
> ------------------------------
>
> In daliuge-engine/dlg/apps/pyfunc.py
> <https://github.com/ICRAR/daliuge/pull/346#discussion_r2196944941>:
>
> > @@ -667,17 +669,21 @@ def _ports2args(self, pargsDict, keyargsDict) -> dict:
>                  key = list(inport.keys())[0]
>                  inputs_dict[key] = {"name": inport[key], "path": None, "drop":
>                      self._inputs[key]}
> +            skip_on_input = True
>
> I don't believe this is necessary any more, as we don't use it in the
> identify_named_ports function.
> ------------------------------
>
> In daliuge-engine/dlg/drop.py
> <https://github.com/ICRAR/daliuge/pull/346#discussion_r2196949996>:
>
> > +                c_param = kwargs.get(attr_name)
>              elif has_app_param:
>
> May we call these component_param, attribute_param? It's just that much
> more readable.
> ------------------------------
>
> In daliuge-engine/dlg/drop.py
> <https://github.com/ICRAR/daliuge/pull/346#discussion_r2196950774>:
>
> >                      "InputPort",
>                      "OutputPort",
>                      "InputOutput",
>                  ]:
> -                    # inp = kwargs["input"]
> -                    # param = pickle.loads(
> -                    #     droputils.allDropContents()
> -                    #     )
> -                    pass
> -                else:
> -                    param = kwargs["applicationArgs"].get(attr_name).value
> -            return param
> +                    a_param = kwargs["applicationArgs"].get(attr_name).value
> +            if has_component_param:
> +                # logger.warning(
>
> There's quite a few comments that exist that we can delete now that
> testing is done.
> ------------------------------
>
> In daliuge-engine/dlg/drop.py
> <https://github.com/ICRAR/daliuge/pull/346#discussion_r2196953702>:
>
> > -                    #     )
> -                    pass
> -                else:
> -                    param = kwargs["applicationArgs"].get(attr_name).value
> -            return param
> +                    a_param = kwargs["applicationArgs"].get(attr_name).value
> +            if has_component_param:
> +                # logger.warning(
> +                #     "Drop has both component and app param %s. Using component param.",
> +                #     attr_name
> +                # )
> +                return c_param
> +            elif has_app_param:
> +                return a_param
> +            else:
> +                return default_value
>
> Aren't we having some scope creep between our definitions of component vs
> attribute here?
> ------------------------------
>
> In daliuge-engine/dlg/drop.py
> <https://github.com/ICRAR/daliuge/pull/346#discussion_r2196956311>:
>
> >                      "InputPort",
>                      "OutputPort",
>                      "InputOutput",
>                  ]:
> -                    # inp = kwargs["input"]
> -                    # param = pickle.loads(
> -                    #     droputils.allDropContents()
> -                    #     )
> -                    pass
> -                else:
> -                    param = kwargs["applicationArgs"].get(attr_name).value
> -            return param
> +                    a_param = kwargs["applicationArgs"].get(attr_name).value
> +            if has_component_param:
> +                # logger.warning(
>
> This is the second check for has_component_param. We don't need the
> variables c_param, a_param, we should be able to return
> kwargs.get(attr_name) immediately.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ICRAR/daliuge/pull/346#pullrequestreview-3004460245>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJVMSA5Z4YTNGH5NT5AU4L3HYO2TAVCNFSM6AAAAACAWGXZTGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTAMBUGQ3DAMRUGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

myxie commented:
Thank you for picking up this work @awicenec and addressing the PR feedback. I'm going to merge this now so we can release it as a patch-fix, and will raise some JIRA tickets to address some of the existing technical debt we have worked around in this PR.
